### ISSUE TITLE
Where is the script "pull-and-make-glasgow-solver.sh?

### ISSUE THREAD

**Issue by lichengzhang1** [markirch/sat-modulo-symmetries#21]
Hello,
I see that if I need to set up the exclusion of subgraphs myself, I should execute the script "pull-and-make-glasgow-solver.sh". But it seems I haven't found it in the directory.

Where is the script "pull-and-make-glasgow-solver.sh? the script `pull-and-make-glasgow-solver.sh` 

Thanks.
Licheng.

**Comment by peitl** [comment]
You are right, this is a bug, I'll fix it asap. I suspect that you cannot use `build-and-install.sh` at all if you want Glasgow (you have to do all the CMake steps manually), but I'll try to fix that as well.

**Comment by peitl** [comment]
I have just pushed an update to `build-and-install.sh`, please see if it works now. It should pull and build the Glasgow solver when you pass `-s` (see `-h` for help). You may need to clear the `build` directory first.

**Comment by lichengzhang1** [comment]
It works. I appreciate your help.  By the way,  is it possible to provide such a similar option in Python swap as well?  I just see the option ` --forbidden-subgraphs -flie `is offered in ./smsg, not in `graph_builder.py`.

For example, if I would like to calculate the Planar Turan numbers for $P_5$-free planar graphs  ($P_5$ is a path with 5 vertices).  Did it be done? Or where should we consider adding an interface for this? This would be more flexibility for user.  

Best wishes,
Licheng.

**Comment by peitl** [comment]
You can pass any arguments to `smsg` through the wrapper using `--args-SMS="--forbidden-subgraphs file"`, but I agree that this deserves a better interface.

Note that at the moment it is only possible to block *non-induced* subgraphs. The Glasgow solver can also do induced, but there is currently no way to pass that through our interface. I assume for $P_5$-free, you want induced $P_5$? I should be able to add this soon-ish.

**Comment by lichengzhang1** [comment]
Yes, I currently need it. (yes, induced $P_5$ in my ideas) Thank you. And I'm more interested in learning how to add it. Of course, I'm very looking forward to more  arguments in your nice programs.

**Comment by manfredscheucher** [comment]
Concerning "how to add it", the `graph_builder.py` python interface is quite simple and hopefully self-explaining: 
1. To register a new command line parameter, you just start looking at `getDefaultParser`. similar to the other parameters, you could add something like 
```
solve_args.add_argument("--forbidden-subgraphs", type=str, help="write some help text, some info about the file to be passed"
```
The parsing is handled by the python ArgumentParser [see this link](https://docs.python.org/3/library/argparse.html)
Note that we split up the arguments into 3 groups main_args,solve_args, and constraint_args, but this is not strict. So if you would want to add say 10 new parameters for your functionality you could make another group

2. essentially you can use Ctrl+F to see what happens with other parameters (the program is very simple and rather straight forward). note that parameters with hyphen symbol like `args-SMS` are refered to as `args.args_SMS` because `-` (minus symbol) is converted to `_` (underline symbol) by the ArgumentParser. So essentially, you could add the ```args.args_SMS += f" --forbidden-subgraphs {args.forbidden_subgraphs}"``` in the souce, so that you can simply use `--forbidden-subgraphs file` for the python interface instead of the rather long command `--args-SMS="--forbidden-subgraphs file"`, however, this makes it less transparent that it is actually a command which is passed, so we decided not to do it for others

**Comment by peitl** [comment]
@lichengzhang1 I just pushed an update with which you can block induced subgraphs, with `--forbidden-induced-subgraphs FILE` (analogously to forbidden non-induced subgraphs). This is still experimental, let me know if you find any problems.

**Comment by lichengzhang1** [comment]
@peitl  Great. I have verified $P_5$-free (induced)  graphs with another program, and your results match numerically, very nice.  By the way, does smsg provide any options for restricting the number of edges?( I discussed your program with Professor Brendan McKay who wrote `nauty`, and he expressed very interest, but he also found it unusual that there are many differences between the options for graph_bulider.py , smsg and smsd. I told him some of your explanations.)

**Comment by lichengzhang1** [comment]
@manfredscheucher  Thanks. Now it works.

**Comment by peitl** [comment]
@lichengzhang1 this should be the wrapper arguments `--num-edges-upp` and `--num-edges-low`
